cardano-api/src/Cardano/Api/IO.hs

```diff
@@ -0,0 +1,41 @@
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE InstanceSigs #-}
+{-# LANGUAGE KindSignatures #-}
+
+module Cardano.Api.IO
+  ( File(..)
+  , In
+  , Out
+  , InOut
+  , HasFileMode(..)
+  ) where
+
+import           Data.Kind (Type)
+import           Data.String (IsString)
+
+-- | Phantom type to indicate a file for reading.
+data In
+
+-- | Phantom type to indicate a file for writing.
+data Out
+
+-- | Phantom type to indicate a file for reading or writing.
+data InOut
+
+-- | A file path with additional type information to indicate whether it is to
+-- be used for reading or writing.
+newtype File direction = File { unFile :: FilePath }
+  deriving (Eq, Ord)
+  deriving newtype (Show, IsString)
+
+class HasFileMode (f :: Type -> Type) where
+  usingIn :: f InOut -> f In
+  usingOut :: f InOut -> f Out
```

Comment:
Some functions will only take either `File In` or `File Out`.  A `File InOut` is not directly usable by such functions, but it can be converted to a `File In` or `File Out` as needed using `HasFileMode` methods.